### PR TITLE
fix(billing): weekly reset semantics + lazy seed + plan-change reset

### LINKED ISSUE #3569: bug(billing): weekly meter reset no-op + first-run blocked when no BillingUsage doc
## Audit findings (2 × CRITICAL — blocks LIVE launch)

Two compounding bugs from the compute-pricing audit. Combined: a new user on any plan cannot scrape, and active subscribers see weekly bucket rollover effectively no-op (4-week spread instead of 4 separate weeks).

### Finding #1 — Weekly reset anchored on `currentPeriodStart` (monthly)

`modules/billing/services/billing.reset.service.js:43, 102-115`

```js
const newWeekKey = isoWeekKey(periodStart);
// ...
await resetWeek(String(sub.organization), new Date(sub.currentPeriodStart));
```

Stripe `current_period_start` only advances at monthly/annual renewal. Within a cycle the cron derives the same `weekKey` every week → `archiveOtherWeeks` archives nothing + `upsertWeekSnapshot` finds the existing doc → reset is no-op for weeks 2/3/4.

**Effect:** weekly bucket = monthly bucket in practice. Pro user (quota 200k/wk) consumes 200k spread over 4 weeks instead of 4 × 200k.

### Finding #2 — `requireQuota` blocks before first attribution

`modules/billing/middlewares/billing.requireQuota.js:72-80` reads quota from BillingUsage doc:

```js
const usage = await BillingUsageService.getMeter(orgId);
const meterQuota = usage?.meterQuota ?? 0;
if ((meterQuota - meterUsed) + extrasBalance <= 0) return 402;
```

`getMeter` (`billing.usage.service.js:225-228`) returns `null` when no doc exists for the current week. The doc is only created on `incrementMeter` — which runs **after** the scrap completes. Middleware runs **before** → deadlock.

**Effect:** any new user (incl. Free with quota=10 in migration) returns 402 on the first scrap of the week.

## Proposed fix

1. **Lazy-fallback to BillingPlan in middleware** when no BillingUsage doc:
   ```js
   if (!usage) {
     const sub = await SubscriptionRepository.findPlan(orgId);
     const plan = await BillingPlanService.getActivePlan(sub?.plan ?? defaultPlan);
     meterQuota = plan?.meterQuota ?? 0;
     meterUsed = 0;
   }
   ```
   Don't write the doc — let `incrementMeter` do it on first attribution.

2. **Fix `resetAllDue`** to derive the week anchor from `lastResetAt + 7d` (or `now`), not `currentPeriodStart`. Period_start only seeds initial subscription; weekly rollovers are local time math.

## Tests required

- New Free user: 1st scrap passes (currently 402)
- Subscription at S+1/S+2/S+3 in same Stripe cycle: bucket actually advances
- Attribution at S+1 increments the new bucket, not the previous week's
- `archiveOtherWeeks` correctly archives prior weekKey (regression)

## Files

- `modules/billing/middlewares/billing.requireQuota.js:42-90`
- `modules/billing/services/billing.usage.service.js:100-228`
- `modules/billing/services/billing.reset.service.js:43-91, 102-123`
- `modules/billing/repositories/billing.subscription.repository.js:156-170`

## Audit context
Source: codex billing audit (session `019de402-8cea-7cf2-b628-4663071a9d5b`). Independently verified.

### LINKED ISSUE #3571: bug(billing): plan upgrade mid-cycle does not trigger resetWeek
## Audit finding (HIGH — blocks LIVE launch, violates memo)

Memo edge case 5: \"Plan upgrade mid-cycle = switch immédiat + reset compteur hebdo + extras conservés\".

Current implementation only triggers `resetWeek` when `current_period_start` changes. Stripe does NOT change `current_period_start` on plan switch (proration only) → upgrade/downgrade mid-cycle leaves the meter bucket untouched.

### Evidence

`modules/billing/services/billing.webhook.service.js:246-258`:
```js
const previousPeriodStart = event?.data?.previous_attributes?.current_period_start;
if (
  previousPeriodStart !== undefined &&
  subscription.current_period_start !== previousPeriodStart &&
  newPeriodStart
) {
  await BillingResetService.resetWeek(organizationId, newPeriodStart);
}
```

The `plan.changed` event is emitted at lines 222-244 but does NOT trigger any reset.

### Concrete scenario

User on Pro (quota 200k/wk) consumed 180k. Downgrades to Growth (quota 50k/wk) mid-cycle. Webhook fires `customer.subscription.updated` with new plan but same period_start.
- `resetWeek` not called
- `requireQuota` reads `meterUsed=180k`, `meterQuota=50k` (fresh from new plan via planVersion bump)
- `remaining = 50k - 180k + 0 = -130k` → 402 BLOCK until end of period

Symmetric on upgrade: user on Growth (50k) maxed out upgrades to Pro (200k). Without reset, `meterUsed=50k`, `meterQuota=200k` → only 150k available the rest of the week instead of full 200k.

## Proposed fix

In `handleSubscriptionUpdated`, after detecting `previousPlan !== newPlan`, trigger `resetWeek` even if period_start unchanged. Use current period_start as anchor (or `now` to align with the lazy-seed fix from #3569).

```js
if (previousPlan && previousPlan !== newPlan) {
  // ... existing event emit
  try {
    const anchor = newPeriodStart ?? new Date(); // fallback now if period unchanged
    await BillingResetService.resetWeek(organizationId, anchor);
  } catch (err) {
    console.error('[billing.webhook] resetWeek on plan change failed (non-fatal):', err?.message);
  }
}
```

## Tests required

- subscription.updated with plan change, same period_start: resetWeek called once
- subscription.updated with plan change AND period_start change: resetWeek called once (not twice)
- subscription.updated without plan change: resetWeek NOT called
- After upgrade Growth→Pro, requireQuota allows fresh full Pro quota
- After downgrade Pro→Growth, meterUsed reset to 0 (extras preserved per memo)

## Files

- `modules/billing/services/billing.webhook.service.js:198-260`

## Audit context
Source: codex billing audit. Verified against memo edge case 5.

### PR DESCRIPTION
**Pull request by PierreBrisorgueil**
## Summary

- **Lazy seed in requireQuota** (`billing.requireQuota.js`): when `getMeter()` returns `null` (new user / first scrap of the week), fall back to `BillingPlanService.getActivePlan()` for `meterQuota` instead of defaulting to 0. Prevents 402 METER_EXHAUSTED on every first request. Reuses the `subscription` already fetched by the degraded-mode gate — no extra DB round-trip.
- **Week anchor fix in resetAllDue** (`billing.reset.service.js`): derive anchor from `lastResetAt + 7d` (or `now` on first run) instead of `currentPeriodStart`. `currentPeriodStart` is constant within a monthly Stripe cycle, so weeks 2/3/4 were resetting to the same `weekKey` → no-op. The new logic advances by exactly 7 days each cron run.
- **Plan-change reset in handleSubscriptionUpdated** (`billing.webhook.service.js`): when `previous_attributes.items` shows a plan change, trigger `resetWeek` immediately (memo edge case 5: plan switch = immediate meter reset + fresh quota). `planChangeResetTriggered` flag prevents double reset when plan and period_start both change simultaneously.

Closes #3569
Closes #3571

### CODE REVIEW

File: modules/billing/middlewares/billing.requireQuota.js [hunk 1]
```diff
@@ -72,8 +73,23 @@ function requireQuota(resource, action) {
...
+          // Fall back to the plan quota so first-run requests are not blocked.
+          // Reuse the `subscription` already fetched by the degraded-mode gate above.
+          const planId = subscription?.plan ?? config?.billing?.defaultPlan ?? 'free';
+          const activePlan = await BillingPlanService.getActivePlan(planId);
+          meterUsed = 0;
+          meterQuota = activePlan?.meterQuota ?? 0;
+        } else {
+          meterUsed = usage.meterUsed ?? 0;
+          meterQuota = usage.meterQuota ?? 0;
+        }
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/middlewares/billing.requireQuota.js:101]: Acknowledged as informational. The plan-change webhook (`handleSubscriptionUpdated`) calls `resetWeek` on plan switches, which upserts a new `BillingUsage` doc with the updated `meterQuota` snapshotted from the active plan. The stale-quota window is bounded by webhook latency. Accepted design trade-off per current architecture.

File: modules/billing/services/billing.reset.service.js [hunk 2]
```diff
@@ -110,7 +110,13 @@ const resetAllDue = async () => {
 
   for (const sub of subs) {
     try {
-      await resetWeek(String(sub.organization), new Date(sub.currentPeriodStart));
+      // Derive the week anchor from lastResetAt + 7d (or now when no prior reset exists).
+      // Using currentPeriodStart would derive the same weekKey every run within the same
+      // monthly/annual Stripe cycle → reset would be a no-op for weeks 2/3/4.
+      const anchor = sub.lastResetAt
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/services/billing.reset.service.js:119]: Fixed in 032c92eb: `resetAllDue` now uses `Math.max(lastResetAt+7d, now)` to clamp the anchor to the current moment when the cron runs late. This ensures the reset always targets the current ISO week rather than a stale past bucket.

File: modules/billing/middlewares/billing.requireQuota.js [hunk 3]
```diff
@@ -72,8 +73,23 @@ function requireQuota(resource, action) {
         const usage = await BillingUsageService.getMeter(orgId);
         const extrasBalance = await BillingExtraBalanceRepository.getBalance(orgId);
 
-        const meterUsed = usage?.meterUsed ?? 0;
-        const meterQuota = usage?.meterQuota ?? 0;
+        let meterUsed;
+        let meterQuota;
+
+        if (!usage) {
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/middlewares/billing.requireQuota.js:79]: Already implemented in commit cd6a5738 (fix #3569). The suggestion matches the current code exactly: when `usage` is null, we fetch `activePlan` and use `activePlan.meterQuota`. Additionally, in 0f917f4e, we now also guard against `activePlan === null` returning 503 PLAN_NOT_CONFIGURED instead of silently defaulting to meterQuota=0.

File: modules/billing/tests/billing.webhook.subscription.unit.tests.js [hunk 4]
```diff
@@ -187,6 +187,224 @@ describe('Billing webhook subscription unit tests:', () => {
...
+            },
+          },
+        },
+      );
+
+      expect(mockResetService.resetWeek).toHaveBeenCalledTimes(1);
+      expect(mockResetService.resetWeek).toHaveBeenCalledWith(
+        orgId,
+        new Date(periodStart * 1000),
+      );
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/tests/billing.webhook.subscription.unit.tests.js]: Fixed in 032c92eb. The test was updated: for a plan-only change (same period_start), the test now asserts `expect(mockResetService.resetWeek).toHaveBeenCalledWith(orgId, expect.any(Date))` and additionally verifies the anchor is NOT the billing-cycle period start. This thread is now outdated.

File: modules/billing/tests/billing.reset.service.unit.tests.js [hunk 5]
```diff
@@ -330,5 +330,97 @@ describe('BillingResetService unit tests:', () => {
...
+        { lastResetAt: new Date('2026-05-01T12:00:00.000Z'), expectedKey: '2026-W19' }, // +7d = May 8
+        { lastResetAt: new Date('2026-05-08T12:00:00.000Z'), expectedKey: '2026-W20' }, // +7d = May 15
+      ];
+
+      for (const { lastResetAt, expectedKey } of runsAndExpectedKeys) {
+        jest.resetModules();
+        const anchor = new Date(lastResetAt.getTime() + 7 * 24 * 60 * 60 * 1000);
+        jest.setSystemTime(anchor);
+
+        // Re-import with fresh mocks for each simulated run
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/tests/billing.reset.service.unit.tests.js]: Fixed in 0f917f4e: removed the dead `jest.resetModules()` from the loop and updated the comment from 'Re-import with fresh mocks for each simulated run' to 'Reconfigure mocks for each simulated run'. The module is not re-imported inside the loop, so resetModules() was adding overhead without effect.

File: modules/billing/services/billing.webhook.service.js [hunk 6]
```diff
@@ -240,12 +241,26 @@ const handleSubscriptionUpdated = async (subscription, event) => {
...
       }
+
+      // Memo edge case 5: plan switch mid-cycle = immediate reset of weekly meter.
+      // Stripe does NOT advance current_period_start on plan switch (proration only),
+      // so we must trigger resetWeek here independently of the period-start change block.
+      // Use newPeriodStart when available (renewal happened simultaneously), fall back to now.
+      const anchor = newPeriodStart ?? new Date();
+      try {
+        await BillingResetService.resetWeek(organizationId, anchor);
+        planChangeResetTriggered = true;
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/services/billing.webhook.service.js:263]: Fixed in 032c92eb. Added `periodChanged` flag: `const periodChanged = previousPeriodStart !== undefined && subscription.current_period_start !== previousPeriodStart && newPeriodStart`. Anchor is now `new Date()` for plan-only mid-cycle switches, and `newPeriodStart` only when the billing period also changed simultaneously. This correctly targets the current ISO week.

File: modules/billing/middlewares/billing.requireQuota.js [hunk 7]
```diff
@@ -72,8 +73,23 @@ function requireQuota(resource, action) {
...
+          // No BillingUsage doc yet (new user / first scrap of the week).
+          // Don't create the doc here — let incrementMeter do it on first attribution.
+          // Fall back to the plan quota so first-run requests are not blocked.
+          // Reuse the `subscription` already fetched by the degraded-mode gate above.
+          const planId = subscription?.plan ?? config?.billing?.defaultPlan ?? 'free';
+          const activePlan = await BillingPlanService.getActivePlan(planId);
+          meterUsed = 0;
+          meterQuota = activePlan?.meterQuota ?? 0;
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/middlewares/billing.requireQuota.js]: Fixed in 0f917f4e: added explicit guard for `activePlan === null || activePlan === undefined` returning 503 PLAN_NOT_CONFIGURED instead of falling through to `meterQuota=0` (which would have surfaced as 402 METER_EXHAUSTED). Also added a test: 'fix #3575: no meter doc + getActivePlan returns null — returns 503 PLAN_NOT_CONFIGURED'.

File: modules/billing/services/billing.reset.service.js [hunk 8]
```diff
@@ -110,7 +110,13 @@ const resetAllDue = async () => {
...
+        ? new Date(new Date(sub.lastResetAt).getTime() + 7 * 24 * 60 * 60 * 1000)
+        : now;
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/services/billing.reset.service.js:121]: Fixed in 032c92eb: `resetAllDue` now uses `Math.max(new Date(sub.lastResetAt).getTime() + 7 * 24 * 60 * 60 * 1000, now.getTime())` which is exactly the suggestion. When the cron is delayed >1 week, the anchor clamps to `now` ensuring the reset targets the current ISO week.

**Comment by PierreBrisorgueil** [comment]
@coderabbitai full review
